tests/onlinequantile/main.c

```diff
@@ -0,0 +1,173 @@
+#include <stdio.h>
+#include <string.h>
+
+#include "onlinequantile.h"
+#include "parsenum.h"
+#include "warnp.h"
+
+#define LINELEN 80
+
+/* Return 0 if ok, 1 on EOF, -1 on error. */
+static int
+parse_line(FILE * fp, char * instr, double * val)
+{
+	char line[LINELEN];
+
+	/* Repeat until we have a non-comment, non-blank line. */
+	do {
+		if (fgets(line, LINELEN, fp) == NULL) {
+			/* Handle error or EOF. */
+			if (ferror(fp)) {
+				warnp("fgets");
+				goto err0;
+			}
+			goto eof;
+		}
+	} while ((line[0] == '#') || (strlen(line) == 1));
+
+	/* Sanity-check the format.  Multiple tabs characters are ok. */
+	if ((line[1] != '\t') || (line[strlen(line) - 1] != '\n')) {
+		warn0("Unexpected character in line: %s", line);
+		goto err0;
+	}
+
+	/* Remove newline. */
+	line[strlen(line) - 1] = '\0';
+
+	/* Get instruction and value from the line. */
+	*instr = line[0];
+	if (PARSENUM(val, &line[2])) {
+		warnp("parsenum");
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+eof:
+	/* End of file! */
+	return (-2);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+static int
+handle_line(struct onlinequantile * oq, char instr, double val)
+{
+	double x;
+
+	switch (instr) {
+	case 'a':
+		/* Add value from file. */
+		if (onlinequantile_add(oq, val)) {
+			warn0("onlinequantile_add");
+			goto err0;
+		}
+		break;
+	case 'g':
+		/* Get the online quantile, and check against expected value. */
+		if (onlinequantile_get(oq, &x)) {
+			warn0("onlinequantile_get");
+			goto err0;
+		}
+		if (fabs(x - val) >= 0.001) {
+			warn0("unexpected value: %g instead of %g", x, val);
+			goto err0;
+		}
+		break;
+	default:
+		warn0("Unexpected instruction: %c", instr);
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+static int
+process_test(const char * filename)
+{
+	struct onlinequantile * oq;
+	FILE * fp;
+	char instr;
+	double val;
+
+	/* Open file. */
+	if ((fp = fopen(filename, "r")) == NULL) {
+		warnp("fopen");
+		goto err0;
+	}
+
+	/* Get first line and initialize with the desired quantile. */
+	if (parse_line(fp, &instr, &val)) {
+		warnp("parse_line");
+		goto err1;
+	}
+	if ((oq = onlinequantile_init(val)) == NULL) {
```

Comment:
If you're going to do this you should probably check that `instr == 'q'`.

Alternatively you could handle the `q` case in `handle_line`.